README.md
# Lambda Complex

Lambda Complex is a Node.js framework for applications that run entirely within
Lambda, SQS, and other high abstraction layer AWS services. It is well suited to
non-realtime content generation and other types of application driven by message
queues and which require enforced concurrency limits. Examples include high
volume generation of static content from data or other types of workflow
initiated in response to messages placed into SQS queues.

  * [In a Nutshell](#in-a-nutshell)
  * [Why Lambda Complex?](#why-lambda-complex)
  * [Examples](#examples)
  * [How Does Lambda Complex Compare With JAWS?](#how-does-lambda-complex-compare-with-jaws)
  * [Basic Concepts](#basic-concepts)
  * [Creating a Lambda Complex Application](#creating-a-lambda-complex-application)
  * [Working With a Lambda Complex Application](#working-with-a-lambda-complex-application)
  * [A Speculative Roadmap](#a-speculative-roadmap)

## In a Nutshell

The high points of Lambda Complex:

* Assemble applications from Node.js Lambda function implementations. Use any
NPM module that exports one or more Lambda function handlers, or write your own.
* Configure the Lambda Complex application to pass data between Lambda
function invocations in potentially complicated, result-dependent ways.
* Each Lambda function in the application can be linked to an SQS queue, and is
then invoked in response to messages sent to that queue.
* Currently this is the only way to send data to a Lambda Complex application.
* A Lambda complex application is deployed as a CloudFormation stack.

A Lambda Complex application can be deployed from a developer machine, or from
an Amazon Linux EC2 instance [if binary NPM modules are required][5]. No server
infrastructure beyond that is needed.

The typical Lambda Complex application consists of a few small Node.js packages
exposing Lambda function handlers, and a configuration file to define necessary
details such as permissions to access AWS resources. When deployed, the Lambda
functions interact with other AWS services to, for example, generate content in
response to SQS messages, or process a workflow in many small steps.

## Why Lambda Complex?

The name? Because it involves AWS Lambda and it is a little bit complicated.

Joining together many Lambda functions to form an application might sound a
little like reinventing the wheel, given the existence of AWS Simple Workflow
Service and other similar frameworks. Why do this, especially given the work
needed to factor any complex application into functions that will run within the
time, memory, thread, and other constraints of AWS Lambda?

The answer is that this is largely a way to flee from devops requirements in
more traditional server deployments. A Lambda Complex application requires no
servers and thus no complicated deployment infrastructure. No provisioning
scripts, no upgrade treadmill, no version conflicts: this particular combination
of AWS services abstracts away all of that. Just design the application, write
the code, write the configuration file, run the deployment script, and you are
done and deployed, your application running.

## Examples

You should find the following examples informative if looking for details, or
help in constructing your own application:

* An [example application configuration file][1] with documentation.
* A [simple example application][2].

## How Does Lambda Complex Compare With JAWS?

[JAWS][6] is great. Lambda Complex achieves much the same outcome, which is to
say constructing an application that resides entirely in AWS Lambda and related
AWS services. The two are different in near every design choice at the detail
level, however.

At the present time Lambda Complex is more suited to:

* Slower, non-time-critical applications, such as static content generation.
* Applications that make use of existing third party Node.js Lambda functions.
* Applications accepting data via SQS queues.
* People who like monitoring application state via SQS queue attributes.
* People who don't mind a lack of structure in their frameworks.

JAWS is more suited to:

* Faster applications, such as APIs.
* Applications that must use non-Node.js Lambda functions.
* Applications accepting data through the standard Lambda function channels.
* People who like a little more structure in their frameworks.

## Basic Concepts

![Lambda Complex overview][9]

### Application Configuration File

A Lambda Complex application is defined by a configuration file that specifies
the AWS details, the components and their Lambda function packages, the IAM
roles to grant access to AWS resources, and other necessary items.

### Component

A Lambda Complex component consists of:

* A Lambda function implementation.
* An SQS queue to measure concurrency.
* A method of passing data to the Lambda function data, such as an SQS queue.
* The component definition in the Lambda Complex application configuration file.

The following types of component exist:

#### Event From Message

The component is bound to a specific queue, and when invoked will consume a
message from that queue to provide the event data passed to the handler.

This type of component has a specified concurrency, and is limited to that
number of concurrent invocations when the queue is populated. The coordinator
periodically invokes instances of the component Lambda function when the queue
has visible messages.

If the component fails in its operation, the message remains in the queue and
will be retried. Therefore this type of Lambda function can be written to fail
fast and with no intricate error handling.

#### Event From Invocation

The component is a normal Lambda function, and event data is passed to the
handler in the normal way. The component Lambda function is invoked by other
component Lambda functions directly, without intervention of the coordinator.

The advantage of this is speed, not having to wait for a queue check, while the
disadvantage that this component must be absolutely bulletproof or data and
operations will be lost when it fails.

#### Internal

Internal components are built-in Lambda functions that manage the operation of
a Lambda Complex application, such as responding to queues backlogs.

### Lambda Complex Code Wraps Provided Lambda Function Implementations

Lambda Complex applications can use arbitrary third party Lambda function
handler implementations, so long as they are Node.js and have a `package.json`
file, because during deployment those implementations are wrapped in Lambda
Complex code.

When the component Lambda function is invoked in a deployed Lambda Complex
application, Lambda Complex code runs first to take care of the necessary
details - such as retrieving data from an SQS queue, tracking concurrency, and
so forth. Only then does it pass control to the underlying Lambda function
handler.

### Coordinator

The coordinator is a built-in Lambda Complex Lambda function that checks the
state of the running application and invokes Lambda functions in response to
circumstances. E.g. on the arrival of new messages in SQS queues.

### Invoker

The invoker is a built-in Lambda Complex Lambda function that is used to invoke
large numbers of other Lambda functions. API calls take time, and a Lambda
function instance is both time-limited and thread-limited. Therefore invoking
a very large number of Lambda function instances requires the creation of
additional invoker instances to carry out that work.

### Concurrency Queues

Each component is associated with an SQS queue used to measure the number of
concurrent invocations of the component Lambda function. When the component
Lambda function is invoked, a message is posted to the queue. That message is
deleted on completion of the Lambda function, or expires after the Lambda
function timeout.

Note that queue messages cannot expire more rapidly than 60 seconds, and Lambda
functions can have shorter timeouts. This shouldn't be too much of an issue
since Lambda Complex reactions to uncaught exceptions to take the necessary
actions - it is fairly hard to cause the concurrency message deletion to fail
to take place.

### ARN Map

During deployment of the CloudFormation stack for a Lambda Complex application,
a JSON map of the relevant ARNs for queues and Lambda functions is created and
uploaded to S3. This file is necessary for the Lambda Complex framework code to
function, and is loaded by every invoked Lambda function before they take
action.

## Creating a Lambda Complex Application

### Design the Application

Break up your application functionality into small chunks that can run in
parallel, and which individually cannot last longer than the Lambda time limit
(currently 300 seconds). Each of these is a component. Identify the data that
passes from component to component, and points at which flow control decisions
must be made: route to component `A` versus component `B` based on the data.

Not all applications are a good fit for Lambda Complex, and it is worth taking a
little time to figure out whether or not this is a good plan for any specific
use case.

For each component you will need to write or use a third party Lambda function
provided in an NPM package. Plan for a component to accept all of its run-time
configuration options, such as an environment specifier, resource names, and so
on, via the event passed as an argument to the Lambda function handler.

## Create the Application Configuration File

The [example configuration file][1] provides a good starting point, as it
contains examples of nearly all that is needed to specify much larger
applications.

### Specify Deployment Details

Specify the necessary deployment details:

```
{
  name: 'simple',
  version: '0.1.0',
  // Add
  deployId: Math.round((new Date()).getTime() / 1000),

  deployment: {
    region: 'us-east-1',
    s3Bucket: 'lambda-complex',
    s3KeyPrefix: 'applications/',
    // No additional tags.
    tags: {},
    switchoverFunction: function (stackDescription, config, callback) {
      callback();
    }
  },
```

The most important and potentially most complex item here is the
`switchoverFunction`. This is invoked after deployment of a new version of the
application, but prior to deletion of the previous version of the application.
It should be used to make any changes needed to switch resources to use the
new applications.

In a Lambda Complex application, this means the SQS queues that will be used to
introduce data into the application: the new queues will have different
identities, and the applications sending to them should be updated.

### Configure the Coordinator Behavior

The coordinator configuration is as important as the `concurrency` property of
components when tailoring an application to work within the Lambda limits placed
on an AWS account. See the example configuration for an explanation of these
properties, and note that it is perfectly possible to set values that are too
high for a coordinator invocation to complete without error.

```
  coordinator: {
    coordinatorConcurrency: 1,
    maxApiConcurrency: 10,
    maxInvocationCount: 20,
    minInterval: 10
  },
```

### Add Component and Role Definitions

For each component identified in the design, create a definition in the
`components` array. Most components will be of the `eventFromMessage` type and
this associated with an SQS queue.

Each component must have an associated role, applied to its Lambda function.
Add these to the `roles` array.

```
  roles: [
    {
      name: 'default',
      // No extra statements beyond those added by default to access queues.
      statements: []
    }
  ],

  components: [
    {
      name: 'invokedProcessor',
      type: 'eventFromInvocation',
      // Since this defines no routing, this is a dead end: events are delivered
      // here and no further processing results.
      // routing: undefined,
      lambda: {
        npmPackage: path.join(__dirname, 'processor'),
        handler: 'index.handler',
        memorySize: 128,
        timeout: 60,
        role: 'default'
      }
    }
  ],
```

For most components the `routing` property is important: it determines which
other components accept the output of this component as their input. This can be
a component name, array of component names, or a function:

```
    routing: 'aComponent',

    ...

    routing: ['component1', 'component2'],

    ...

    routing: function (error, data) {
      // This is safe provided that the Lambda function for this component
      // includes the underscore module.
      var _ = require('underscore');

      // Don't send on any data if an error resulted.
      if (error) {
        return [];
      }

      // Otherwise split up or manipulate data and send it to other
      // components as desired.
      return [
        { name: 'componentA', data: _.keys(data) },
        { name: 'componentB', data: _.values(data) }
      ];
    },
```

### Write the Lambda Functions

A Lambda function for use in a Lambda Complex application is written in the
normal way. The result it returns to `context.succeed` or `context.done` may be
routed to one or more other components depending on definitions in the
application configuration file.

Thus any suitable NPM module that provides exported handle functions that
conform to the Lambda specification can be used. As a reminder, these functions
must be of the form:

```
exports.fn = function (event, context) {
  var result = {
    name: 'value'
  };

  context.succeed(result);
}
```

A function must invoke the `context` methods on completion to provide data that
can be passed on to other components in the application.

### Ensure Compatibility with the Supported Node.js Version

AWS Lambda runs under a [specific version of Node.js][7]. Set up development and
testing to ensure that your application can function in that environment.

### Set up AWS Credentials

For deployment to work, suitable AWS credentials for the account specified in
the configuration file must be present. The credentials must at a minimum
allow full interaction with CloudFormation stacks, the S3 bucket and key prefix
specified in configuration, CloudWatch logs, Lambda functions, SQS queues, and
roles.

To make credentials available for Lambda Complex, either create a credentials
file in the standard location or set environment variables to hold the key and
secret key. Both options are [described in the AWS SDK documentation][3].

Unlike many Node.js AWS modules, Lambda Complex does not accept IAM access keys
directly in configuration. It is bad practice to specify access keys in code or
configuration for code; you should always add them to the environment in one of
the ways noted above.

## Working With a Lambda Complex Application

### Grunt Tasks

The `lambda-complex` module exports the following [Grunt][8] tasks, which can
be loaded via `grunt.loadNpmTasks('lambda-complex')` in modules that load
`lambda-complex` as a dependency:

  * `lambda-complex-build` - build a specified application.
  * `lambda-complex-deploy` - build and deploy a specified application.

The usage of the tasks is illustrated below; both require the `--config-path`
option to be specified.

### Deploy the Application

Once the Javascript configuration file is ready, deploy the application via a
Grunt task. Note that the `config-path` option can be set to a relative or
absolute path:

```
grunt lambda-complex-deploy --config-path=/path/to/applicationConfig.js
```

Deployment uses the [cloudformation-deploy NPM module][4], so you may want to
take a look at its documentation. In short, the following occurs:

* Request stack creation.
* Wait on the stack status to show show success.
* Run the provided `config.deployment.switchoverFunction`.
* Delete any previous instances of the stack once the switchover is done.

#### Programmatic Interface for Deployment

Deployment can be managed programmatically:

```
var lambdaComplex = require('lambda-complex');
var config = require('/path/to/applicationConfig');

lambdaComplex.deploy(config, function (error, results) {
  if (error) {
    console.error(error);
  }

  console.info(results);
});
```

#### Binary NPM Module Limitations

Some NPM modules build local binaries on installation. If any of the component
Lambda function implementations require binary modules, then the Lambda Complex
application [must be deployed from an Amazon Linux EC2 instance][5] to ensure
that the environments match.

### Building the Application Without Deployment

It is possible to build the application locally without deploying it, which can
be useful during development. This will result in the creation of the zipped
Lambda function packages and CloudFormation template only.

```
grunt lambda-complex-build --config-path=/path/to/applicationConfig.js
```

#### Programmatic Interface for Building

Building without deployment can be managed programmatically:

```
var lambdaComplex = require('lambda-complex');
var config = require('/path/to/applicationConfig');

lambdaComplex.build(config, function (error, results) {
  if (error) {
    console.error(error);
  }

  console.info(results);
});
```

### Shut Down an Application

Since the application includes self-invoking Lambda functions, the only sure
way to shut it down at the present time is to delete the CloudFormation stack.
Deleting the ARN map file uploaded to S3 during application deployment should
also shut things down, as without the list of ARNs for application resources
nothing in Lambda Complex can work.

### Update a Deployed Application

To update the deployed application, simply deploy again. A new stack will be
created and the old stack destroyed when (a) creation is complete and (b) the
`config.deployment.switchoverFunction` calls back.

If you are using SQS queues to pass data into the application, note that this
will require you to update the queue names. This should be accomplished in the
`switchoverFunction` defined in the application configuration, which will be
invoked after creation is successful but before deletion of the prior
application stack.

### Monitor a Running Application

Lambda functions write to CloudWatch Logs and other attributes such as number
of invocations and errors can be monitored via the CloudWatch API. The component
SQS queues can also be monitored via CloudWatch, and the concurrency SQS queues
are especially useful when monitoring application state.

The coordinator SQS concurrency queue should never be empty, for example, as
this indicates that the application has failed and is no longer running.
Component queues used to introduce data to the application would be backing up
at this point as well.

A number of third party services can be used to build monitors and alerts based
on these logs and metrics.

## A Speculative Roadmap

Lambda Complex remains in an early stage of development. Moving forward the
following are intended.

### Switch Coordinators to Use Scheduled Events

When scheduled events for Lambda functions can be set up via CloudFormation,
switch the coordinator instances to run using this functionality rather than via
self-invocation.

### Better Support for Switchover of SQS Queues

The process of switching delivery of data to new queues on each new deployment
of an existing application is painful, and what happens to the messages in the
old queues? They will get dropped as it stands. Helpers to make it easier
to build a suitable switchover function would be useful.

### Monitoring and Log Streaming

A robust monitoring and log streaming module is needed, built on the CloudWatch
APIs. A programmatic interface to how the applications is working (or not
working) is a necessary part of a number of further refinements.

### More Component Types

Add further component types based on other ways to trigger Lambda functions
from AWS resources. E.g. from S3 events or SNS topics.

### Application Testbed

Providing the ability to run an application locally.

### Stress Testing

Tools to stress test a local or deployed application.

[1]: ./examples/exampleApplicationConfig.js
[2]: ./examples/simple
[3]: http://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html
[4]: https://github.com/exratione/cloudformation-deploy
[5]: https://aws.amazon.com/blogs/compute/nodejs-packages-in-lambda/
[6]: https://github.com/jaws-framework/JAWS
[7]: http://docs.aws.amazon.com/lambda/latest/dg/current-supported-versions.html

CHANGELOG.md
# Changelog

## 0.7.0

  * Deal with the changes in NPM since the last update, particularly local
packages now installing as symlinks rather than being copied into place.

## 0.6.0

  * Rename and export Grunt tasks.
  * Update documentation, add an overview diagram.

## 0.5.1

  * Minor configuration validation improvement.

## 0.5.0

  * Improve configuration validation.
  * Expand unit test converage.

## 0.4.0

  * Add option to delete old CloudWatch log groups on deployment.
  * Make deployment options for stack deletion more granular.

## 0.3.2

  * `VisibilityTimeout` cannot be 0 in the `decrementConcurrencyCount` function, as this can cause message deletion to fail silently.

## 0.3.1

  * Alter increment/decrement error behavior in invoker to match coordinator.
  * Fix concurrency decrement for the case in which no message is found.

## 0.3.0

  * Upload `config.js` alongside the other deployment files. Helpful to keep a record.
  * The first coordinator instances following deployment upload `confirm.txt` on success.
  * Deployment waits for `confirm.txt` to exist.

## 0.2.0

  * Add a maximum concurrency limit for `eventFromMessage` components.
  * Allow multiple concurrent coordinator instances.
  * Improve documentation.

## 0.1.0

  * Initial release.


examples/exampleApplicationConfig.js
/**
 * @fileOverview An example configuration file, with documentation.
 *
 * It is very important to note that the configuration object will, after it is
 * loaded by the deployment code, be:
 *
 * - Regenerated into code and added as a file in Lambda function modules.
 * - Loaded in Lambda by all Lambda functions in a Lambda Complex application.
 *
 * This means that almost everything that results in a valid configuration
 * object will just work, but there are things you can do that will result in
 * failure to run in Lambda.
 *
 * In particular note the restrictions on routing functions, described below in
 * the component inline comments.
 */

module.exports = {
  // The application name. It must be unique for a given account.
  //
  // When deploying an application the queue names and Lambda function names
  // are prefixed with the application name.
  name: 'example',

  // The application version. This is used in tags and some names.
  version: '1.0.0',

  // A deployment ID that is unique to this deployment of this application. For
  // example the BUILD_NUMBER value generated by Jenkins tasks, or a unix
  // timestamp, or similar. It can be alphanumeric.
  //
  // This is included in most tags, paths, and names.
  deployId: 15,

  // AWS deployment settings.
  //
  // The application is deployed as a CloudFormation stack, and after successful
  // completion any prior deployed stacks for this application are destroyed. On
  // a failure to deploy the new stack is destroyed.
  deployment: {

    // ----------------------------------------------------------------------
    // Required.
    // ----------------------------------------------------------------------

    // AWS region.
    //
    // This must match whatever region is configured in the deployment
    // environment.
    //
    // TODO: having this and the environment configuration is clumsy. Look at
    // cutting that down to one declaration one way or another.
    region: 'us-east-1',

    // The S3 bucket into which packaged Lambda functions will be uploaded.
    s3Bucket: 'lambda-complex',

    // All packaged Lambda function keys will be prefixed with this string.
    s3KeyPrefix: 'applications/',

    // ----------------------------------------------------------------------
    // Optional.
    // ----------------------------------------------------------------------

    // Any additional tags to add to the CloudFormation stack.
    tags: {
      tagName: 'tagValue'
    },

    // An asynchronous function invoked after the new stack is deployed but
    // before the old stack is destroyed. This should perform any functions
    // needed to switch over to use the new stack and then call back.
    //
    // Generally this will mean switching to use new queues to introduce data
    // into the application and then waiting for the old queues to clear, but
    // the precise details will depend on the application.
    //
    // This function will not be invoked in Lambda functions when deployed, so
    // it can require() and invoke any resources available to Lambda Complex
    // during deployment.
    //
    // @param {Object} stackDescription Stack description, including outputs.
    // @param {Object} config This configuration object.
    // @param {Function} callback Of the form function (error).
    switchoverFunction: function (stackDescription, config, callback) {
      callback();
    },

    // If set true then CloudFormation stacks created in prior deployments of
    // this application are NOT destroyed on a successful deployment.
    //
    // This can be helpful during development.
    skipPriorCloudFormationStackDeletion: false,

    // If set true, then CloudWatch log groups created by the Lambda function
    // instances from prior deployments of this application are NOT destroyed
    // on a successful deployment.
    //
    // CloudWatch log groups are clutter for some people, others want to keep
    // them.
    skipPriorCloudWatchLogGroupsDeletion: false,

    // If set true, then the CloudFormation stack for this deployment is NOT
    // destroyed on deployment failure.
    //
    // This can be helpful during development.
    skipCloudFormationStackDeletionOnFailure: false
  },

  // The coordinator is a lambda function package that is expected to consume
  // this configuration and manage the invocation of invoker lambda functions in
  // the required concurrency.
  coordinator: {
    // The maximum number of concurrent coordinators to run.
    //
    // Each coordinator will carry out its fraction of the overall work to be
    // done. E.g. for 2 concurrent coordinators, each will undertake half of the
    // necessary invocations.
    //
    // More than one coordinator provides a little redundancy against the random
    // slings and arrows of AWS API failure. The application relies on
    // coordinators invoking themselves, and while the coordinator code tries to
    // be bulletproof, there are still a number of points at which failure can
    // occur.
    //
    // So long as one coordinator is running, it will invoke as many other
    // instances as are needed to maintain this specified concurrency.
    coordinatorConcurrency: 2,

    // How many AWS API requests can the coordinator make at any one time?
    // Too many and you'll tend to see errors.
    maxApiConcurrency: 10,

    // How many Lambda functions, excluding the coordinator itself, can a
    // single coordinator instance launch? Too many and it can get timed out.
    // Too few and concurrency won't be maxed out. When making invocation
    // requests, the <maxApiConcurrency> setting is obeyed, so it will limit
    // how many requests are made at one time.
    maxInvocationCount: 50,

    // Minimum interval in seconds between invocations. Has to be equal to
    // or less than the execution time limit for a lambda function, see:
    // http://docs.aws.amazon.com/lambda/latest/dg/limits.html
    //
    // A coordinator may well run longer than this interval before invoking the
    // next coordinator if it has a lot to do.
    //
    // The shorter the interval the more frequently that new component Lambda
    // function instances will be launched in response to messages in their
    // queues. This goes a long way towards determining the pace of the
    // application for small applications and small amounts of data.
    minInterval: 10
  },

  // Every Lambda function, and thus every component, is associated with a
  // single IAM role. This role provides it with all of the necessary
  // permissions to access AWS resources. E.g. S3 buckets, etc.
  //
  // Component definitions each specify the use of one of these roles.
  //
  roles: [
    // In the rare case that a Lambda function needs no permissions, it is
    // possible to assign an empty role. Lambda Complex will add the necessary
    // internal permissions to it, such as queue access and logging.
    {
      name: 'default',
      statements: []
    },
    // More commonly a role will contain one or more statements. These follow
    // the normal form for policy statements inside a CloudFormation role
    // definition (but with lowercase property names, for consistency with the
    // rest of this config file).
    //
    // In addition to the statements provided here, Lambda Complex will add
    // necessary internal permissions, such as queue access and logging.
    {
      name: 's3Read',
      statements: [
        {
          effect: 'Allow',
          action: [
            's3:get*',
            's3:list*'
          ],
          resource: [
            'arn:aws:s3:::exampleBucket1/*',
            'arn:aws:s3:::exampleBucket2/*'
          ]
        }
      ]
    }
  ],

  // A component consists of a lambda function package that will be triggered in
  // response to circumstances, such as a message in a queue or the completion
  // of another Lambda function.
  //
  // There are several different types of component.
  //
  // Any number of components can be defined.
  components: [

    // Example of a lambda function that consumes a message from a queue when
    // invoked, and the message data becomes the event passed to the lambda
    // function handler.
    //
    // This type of component can be run at a given concurrency, controlled by
    // the coordinator.
    {
      // The name of the lambda function definition to be created in
      // AWS.
      name: 'eventFromMessageExample',

      // Type of the component.
      type: 'eventFromMessage',

      // The maximum number of Lambda function instances for this component that
      // will run at any one time. This, coupled with coordinator.minInterval,
      // goes towards determining the rate at which this component's message
      // queue will drain.
      maxConcurrency: 10,

      // Seconds to wait for a message to show up in the queue before giving up.
      // This should not be high as it cuts into short time limit for a Lambda.
      // function. Setting a few seconds may smooth out the operation of an
      // application in future versions of Lambda Complex, but at the present
      // time should make no difference either way.
      queueWaitTime: 5,

      // Define where the results from this Lambda function are sent. The data
      // passed to context.succeed(data) or context.done(null, data) will be
      // send on to another component as the event passed to the handle of its
      // Lambda function.
      //
      // This is an optional property. If not defined, then data is not passed
      // on to any other component.
      //
      // If a string, then results on success will always be sent to an
      // invocation of the specified component's Lambda function. On failure
      // there is no invocation and no data sent.
      routing: 'eventFromInvocationExample',
      //
      // If an array of strings, then a copy of the results on success is sent
      // to an invocation of the Lambda function for each of the specified
      // components.
      //routing: ['componentA', 'componentB'],
      //
      // If a function, then the destination components and the data sent can
      // be specified as desired.
      //
      // Note that this function will be called inside a Lambda function context
      // and so cannot rely on any NPM modules not in that context. Further,
      // the require() statements for resources must be inside the routing
      // function context, as this config file will be loaded in other contexts
      // that do not have those modules available.
      //
      //routing: function (error, data) {
      //  // This is safe provided that the Lambda function for this component
      //  // includes the underscore module.
      //  var _ = require('underscore');
      //
      //  // Don't send on any data if an error resulted.
      //  if (error) {
      //    return [];
      //  }
      //
      //  // Otherwise split up or manipulate data and send it to other
      //  // components as desired.
      //  return [
      //    { name: 'componentA', data: _.keys(data) },
      //    { name: 'componentA', data: _.values(data) }
      //  ];
      //},

      // Detailing the lambda function to be used by this component.
      lambda: {
        // If an NPM package, this is the package name or location for use when
        // installing: npm install <lambdaFunctionPackage>.
        //
        // Note that you can specify a local absolute path instead of a package
        // name. This is the usual methology when constructing an application
        // from multiple components: all of the lambda function definitions are
        // kept locally.
        npmPackage: 'package-name',
        // npmPackage: '/absolute/path/to/local/package/dir',

        // The name of the lambda function handle exported in the package code.
        // This doesn't have to be the same as the name of the lambda function
        // definition.
        handler: 'index.handler',

        // Memory size in MB, with a minimum of 128.
        memorySize: 128,

        // Timeout for the function in seconds.
        timeout: 60,

        // The role used by this function.
        role: 's3Read'
      }
    },

    // Example of a lambda function that is directly invoked with data from a
    // preceding invocation.
    //
    // Note that there is no concept of concurrency here; this is invoked
    // directly by a prior lambda function execution, so operates at the same
    // concurrency.
    {
      // The name of the lambda function definition to be created in
      // AWS.
      name: 'eventFromInvocationExample',

      // Type of the component.
      type: 'eventFromInvocation',

      // Define where the results from this Lambda function are sent. The data
      // passed to context.succeed(data) or context.done(null, data) will be
      // send on to another component as the event passed to the handle of its
      // Lambda function.
      //
      // In this case, no destination is defined so no other function will be
      // invoked as a result of this function completing.
      routing: undefined,

      // Detailing the lambda function to be used by this component.
      lambda: {
        // If an NPM package, this is the package name or location for use when
        // installing: npm install <lambdaFunctionPackage>.
        //
        // Note that you can specify a local absolute path instead of a package
        // name. This is the usual methology when constructing an application
        // from multiple components: all of the lambda function definitions are
        // kept locally.
        npmPackage: 'package-name',
        // npmPackage: '/absolute/path/to/local/package/dir',

        // The lambda function file and function property exported in the
        // package code. For this example that is index.js and the exported
        // property name 'handle'.
        handler: 'index.handler',

        // Memory size in MB, with a minimum of 128.
        memorySize: 128,

        // Timeout for the function in seconds.
        timeout: 60,

        // The role used by this function.
        role: 'default'
      }
    }
  ]

};
